beluga_ros/test/test_point_cloud.cpp

```diff
@@ -0,0 +1,887 @@
+
+    // Fill the PointCloud2 message
+    for (const auto& point : point_data) {
+      *iter_x = point.x();
+      *iter_y = point.y();
+      *iter_z = point.z();
+
+      ++iter_x;
+      ++iter_y;
+      ++iter_z;
+    }
+  }
+
+  // XYZI pointclouds
+  else if (fields == 4) {
+    // Set offset
+    int offset = 0;
+    offset = addPointField(*message, "x", 1, U, offset);
+    offset = addPointField(*message, "y", 1, U, offset);
+    offset = addPointField(*message, "z", 1, U, offset);
+    offset = addPointField(*message, "intensity", 1, U, offset);
+
+    // Set message params
+    message->point_step = static_cast<unsigned int>(offset);
+    message->row_step = message->width * message->point_step;
+    message->is_dense = true;
+    message->data.resize(message->point_step * message->width * message->height);
+
+    // Return empty pointcloud
+    if (empty)
+      return message;
+
+    // Create data iterators
+    beluga_ros::msg::PointCloud2Iterator<T> iter_x(*message, "x");
+    beluga_ros::msg::PointCloud2Iterator<T> iter_y(*message, "y");
+    beluga_ros::msg::PointCloud2Iterator<T> iter_z(*message, "z");
+    beluga_ros::msg::PointCloud2Iterator<T> iter_intensity(*message, "intensity");
+
+    // Fill the PointCloud2 message
+    typename std::vector<T>::size_type i = 0;
+    for (const auto& point : point_data) {
+      *iter_x = point.x();
+      *iter_y = point.y();
+      *iter_z = point.z();
+      *iter_intensity = intensity_data.at(i);
+
+      ++iter_x;
+      ++iter_y;
+      ++iter_z;
+      ++i;
+    }
+  }
+
+  // Error
+  else {
+    throw std::invalid_argument("Number of fields error");
+  }
+  return message;
+}
+
+TEST(TestPointCloud, XYZPointsUnorderedPC) {
+  const auto origin = Sophus::SE3d{};
+  // Define pointcloud params
+  std::vector<float>::size_type fields = 3;
+  unsigned int width = 1;
+  unsigned int height = 5;
+  // Create some raw data for the points
+  // clang-format off
+  const std::vector<Eigen::Vector3f> point_data = {
+    Eigen::Vector3f(1.0F, 2.0F, 3.0F),
+    Eigen::Vector3f(4.0F, 5.0F, 6.0F),
+    Eigen::Vector3f(7.0F, 8.0F, 9.0F),
+    Eigen::Vector3f(10.0F, 11.0F, 12.0F),
+    Eigen::Vector3f(13.0F, 14.0F, 15.0F)
+  };
+  // clang-format on
+  // Create point cloud message
+  const auto message = make_pointcloud<float, beluga_ros::msg::PointFieldF32>(fields, width, height, point_data);
+
+  // Check aligned pointcloud
+  auto cloud = beluga_ros::PointCloud3<beluga_ros::msg::PointFieldF32>(message, origin);
+  auto map = cloud.points();
+  // Check assert
+  for (unsigned i = 0; i < map.cols(); ++i) {
+    ASSERT_EQ(point_data.at(i).x(), map(0, i));
+    ASSERT_EQ(point_data.at(i).y(), map(1, i));
+    ASSERT_EQ(point_data.at(i).z(), map(2, i));
+  }
+
+  // Check sparse pointcloud
+  auto cloud_sparse = beluga_ros::PointCloudSparse3<beluga_ros::msg::PointFieldF32>(message, origin);
+  auto vector = cloud_sparse.points();
```

Comment:
@pvela2017 here and elsewhere, consider materializing the vector like:
```suggestion
  const auto vector = cloud_sparse.points() | ranges::to<std::vector>();
```
Indexing and querying for the cloud points range size implicitly puts additional requirements than those documented.

Reply:
Alright, as per offline chat with @pvela2017, this doesn't work as `rangesv3` deduces a `const Eigen::Map<>` type that deletes its `operator=` overload. Funnily enough, fully specifying the container type like `std::vector<Eigen::Vector<>>` for the implicit conversion to kick in does not work either, but for unrelated reasons -- getting "temporary of non-literal type in a constant expression" errors. We'll leave this as-is, and dig in later.